rfc/rfc-0000-supply-chain-tracing.md

```diff
@@ -0,0 +1,176 @@
+
+After Cartographer matches a `ClusterSupplyChain` to a `Workload`, the resources defined by the supply chain are resolved to a template and template is applied to stamp out a kubernetes resource on the cluster. After the resource is created, outputs located at paths also defined by the template are read and propagated forward to the next resource in the chain. As Cartographer applies this behavior, it can keep a trace for each resource containing:
+
+1. the name in the ClusterSupplyChain
+2. a reference to the template
+3. a reference to the stamped resource
+4. named input resources as defined by the `ClusterSupplyChain`
+5. the name and value of each output resolved from the stamped resource
+
+The status for a Workload or Deliverable should be enhanced to add:
+
+```yaml
+... existing fields ...
+status:
+  ... existing fields ...
+  resources:
+  - name:                  # string
+    templateRef:           # corev1.ObjectReference
+      apiVersion:            # string
+      kind:                  # string
+      namespace:             # string (empty for cluster scoped resources)
+      name:                  # string
+    stampedRef:            # corev1.ObjectReference
+      apiVersion:            # string
+      kind:                  # string
+      namespace:             # string (empty for cluster scoped resources)
+      name:                  # string
+    inputs:
+    - name:                  # string
+    outputs:
+    - name:                  # string
+      value:                 # apiextensionsv1.JSON
+      lastTransitionTime:    # metav1.Time
+```
+
+Based on the [basic supply chain example](https://github.com/vmware-tanzu/cartographer/tree/53402edb8b8914b4cd36ace82da85f83f3daefc1/examples/basic-sc), the Workload status would have the form:
+
+```yaml
+... existing fields ...
+status:
+  ... existing fields ...
+  resources:
+  - name: source-provider
+    stampedRef:
+      apiVersion: source.toolkit.fluxcd.io/v1beta1
+      kind: GitRepository
+      namespace: default
+      name: my-workload
+    templateRef:
+      apiVersion: carto.run/v1
+      kind: ClusterSourceTemplate
+      name: source
+    outputs:
+    - name: url
+      value: http://source-controller.flux-system.svc.cluster.local./gitrepository/default/my-workload/3d42c19a618bb8fc13f72178b8b5e214a2f989c4.tar.gz
+      lastTransitionTime: "2022-02-16T03:29:52Z"
+    - name: revision
+      value: main/3d42c19a618bb8fc13f72178b8b5e214a2f989c4
+      lastTransitionTime: "2022-02-16T03:29:52Z"
+  - name: image-builder
+    stampedRef:
+      apiVersion: kpack.io/v1alpha2
+      kind: Image
+      namespace: default
+      name: my-workload
+    templateRef:
+      apiVersion: carto.run/v1
+      kind: ClusterImageTemplate
+      name: image
+    inputs:
+    - name: source-provider
+    outputs:
+    - name: image
+      value: registry.example/supply-chain/my-workload@sha256:68f8e8fc6e8ede7a411db9182cd695eac7b3e7e19e4ff9dcb9ba21205c135697
```

Comment:
We could switch including the value to use a hash of the value. Clients would be required to fetch the stamped resource in order to resolve the actual value.

Reply:
Switched to a `digest` and `path` instead of a `value` for outputs.

Reply:
I definitely understand the argument about the config being too big, but I think capturing the value here is important. A lot of our modelling assumes that the workload is the only object that a developer needs to care about, now we're saying that they need to go and make these other associations to other objects.

Reply:
I think if we don't have `value` anymore then we'd have to remove `lastTransitionTime` as we won't be storing the old value so we won't know if it's changed. It'd also be misleading as to what `lastTransitionTime`- it could be taken to mean the last time the output path changed.

Reply:
We link to the template already which has the `path`. If we aren't going to include the output values maybe we just remove outputs altogether. But for the record I think this removes a lot of user value so i'm still very much in favour of keeping the output values.

Reply:
I think there is still some utility to including the output path so that developers don't have to go to the templates to find the output path (which may also be hard to decipher as its not necessarily clear that `urlPath` is the output path they're looking for).

I also think there is a lot of value lost by removing `lastTransitionTime`. Is there a way we could keep `value` but just truncate it if it's too long?

Reply:
Can we find a reasonable compromise and keep some kind of short summary in there, at least timestamp and status, so we don't have to access stamped resources to collect that info? 

That would be very helpful and definitely improve the performance when displaying the supply chain topology.

Reply:
lastTransitionTime is still available. I assume by "a short summary" and `status` you mean something on the order of a metav1.Condition#Status and not the full `$.status` from the stamped resource. 

A simple `True` `False` `Unknown` is something we should drive towards, but to do so requires adding semantics that Cartographer is unaware of today and user facing template configuration, which makes it squarely out of scope for this RFC. 

&lt;sarcasim&gt;
If you want, you can assume that all outputs have a status field that just isn't set (meaning unknown) since that would accurately reflect what Cartographer knows today about the stamped resource.
&lt;/sarcasim&gt;

Reply:
> lastTransitionTime is still available

If cartographer isn't storing the values, i'm not sure how it would know if the value has changed.

Reply:
> > lastTransitionTime is still available
> 
> If cartographer isn't storing the values, i'm not sure how it would know if the value has changed.

The `digest` represents the raw value. @cirocosta [raised concerns](https://github.com/vmware-tanzu/cartographer/pull/634#discussion_r812314936) about the stability of the digest if we expect others to use it to verify an output, but it's sufficiently stable for our own purposes.

Reply:
Ah I thought the discussion was heading towards removing `digest`, but if it's sticking around then ya we can definitely still have `lastTransitionTime`.

Reply:
> On RFC 18 @evankanderson raised a valid concern about the potential size of outputs causing issues.

> We could switch including the value to use a hash of the value. Clients would be required to fetch the stamped resource in order to resolve the actual value.

I also thought we were exploring options to address this possible large status size and therefore remove the digest. I will say, it's not clear to me that we have hypothetical supply chains that would create a 1 MB status (or even 500K, assuming that artifact tracing ends up duplicating much of resource tracing). But if we assume that it is, and that we want to add a hash field and remove the digest, I also wonder about @mklanjsek's question:

> Can we find a reasonable compromise and keep some kind of short summary in there, at least timestamp and status, so we don't have to access stamped resources to collect that info?

Options that I can think of:
1. Include a truncated version of the digest, with a character length that is long enough for most urls, revisions, image digests. This would leave only configs as values that need to be retrieved by end users.
2. Rather than include a digest, construct and include the `kubectl get ... | yq . <path>` command that a user could run to get the value. (Of course, there are 2 popular yq tools, so this may cause more problems than it solves)

Reply:
> we want to add a hash field and remove the digest

What do you see as the difference hash and digest? I consider a digest to be the output of a hashing function.

> Include a truncated version of the digest

the key point of a digest is that it's a fixed length (based on the hash algorithm)

> construct and include the `kubectl get ... | yq . <path>` command that a user could run to get the value

We provide the details sufficient to lookup the value with your own set of tools. Generally, I don't expect APIs to make assumptions about the tools that are used to access that API.

Reply:
@sclevine suggested this out-of-band:
> maybe we could add a preview field next to digest that includes, e.g., just the commit SHA, just the image digest, just the first N bytes of the config?
Thoughts on including something like this @scothis?

Reply:
@jwntrs it re-raises the same question about the sensitivity of outputs. For a sufficiently small value of N, it should alleviate all practical concerns around overflowing the size limitation for a k8s resource.

Reply:
@scothis wdyt of something similar to https://github.com/kubernetes/kubernetes/pull/96084 to address the sensitivity?

Reply:
@cirocosta how do we know what value needs to be masked?

Reply:
Can I get some thumbs up/thumbs down on moving forward with the following:

Add a `preview` field with which holds the first 200 bytes of each value.

Reply:
@scothis After today's chat, I'm pretty comfortable with adding the `preview` field with <1000 bytes (200 seems fine to me). If you aren't opposed, would you mind adding it?

Reply:
added